motion/dp_controller/src/quaternion_dp_controller.cpp

```diff
@@ -0,0 +1,136 @@
+    const Eigen::Vector3d &eta_d_pos, const Eigen::Quaterniond &eta_d_ori) {
+
+  // Rotate from inertial/world to body
+  Eigen::Matrix3d R = q.toRotationMatrix();
+  Eigen::Matrix6d K_p = proportionalGainMatrix(R);
+  Eigen::MatrixXd T = Eigen::MatrixXd::Zero(4, 3);
+  T << -q.vec().transpose(),
+      q.w() * Eigen::Matrix3d::Identity() + skew(q.vec());
+  T = 0.5 * T;
+  Eigen::MatrixXd J_inv = Eigen::MatrixXd::Zero(6, 7);
+  J_inv << R.transpose(), Eigen::MatrixXd::Zero(3, 4), Eigen::Matrix3d::Zero(),
+      4 * T.transpose();
+  Eigen::Vector6d nu_tilde = Eigen::Vector6d::Zero();
+  Eigen::Vector6d remove_ori = Eigen::Vector6d::Zero();
+  remove_ori << 1, 1, 1, 0, 0, 0;
+  nu_tilde = nu - nu_d.cwiseProduct(remove_ori);
+
+  // Error Vector
+  Eigen::Vector6d z = errorVector(x, eta_d_pos, q, eta_d_ori);
+
+  // Integral (change Antiwindup to a more advanced one)
+  double maxPosGain = 0.5;
+  double maxAttGain = 0.05;
+  Eigen::Vector6d IntegralAntiWindup = Eigen::Vector6d::Zero();
+  IntegralAntiWindup << maxPosGain, maxPosGain, maxPosGain, maxAttGain,
+      maxAttGain, maxAttGain;
+  integral += m_i_gain * z;
+  integral =
+      integral.cwiseMin(IntegralAntiWindup).cwiseMax(-IntegralAntiWindup);
+
+  Eigen::Matrix6d m_K_d = Eigen::Matrix6d::Zero();
+  m_K_d.diagonal() << m_d_gain;
+  Eigen::Vector6d g = QuaternionPIDController::restoringForceVector(R);
+
+  // gain
+  Eigen::Vector6d gain = -m_K_d * nu_tilde - K_p * z + g;
+  Eigen::Vector6d scale_g = Eigen::Vector6d::Zero();
+  scale_g << 0.9, 0.9, 0.9, 0.3, 0.3, 0.3;
+  gain = -m_K_d * nu_tilde - K_p * z + g.cwiseProduct(scale_g) - integral;
+
+  //------ Debug ----
+  P_debug = K_p * z;
+  I_debug = integral;
+  D_debug = m_K_d * nu_tilde;
+  //-----------------
+
+  // Rounding gain to remove super small values
+  int num_decimals = 3;
+  gain = (gain * pow(10, num_decimals)).array().round() / pow(10, num_decimals);
+  return (Eigen::Vector6d() << gain).finished();
+}
+
+Eigen::Vector6d
+QuaternionPIDController::restoringForceVector(const Eigen::Matrix3d R) {
+  Eigen::Vector3d f_G = R.transpose() * Eigen::Vector3d(0, 0, m_W);
+  Eigen::Vector3d f_B = R.transpose() * Eigen::Vector3d(0, 0, -m_B);
+  Eigen::Vector6d g = Eigen::Vector6d::Zero();
+  return (Eigen::Vector6d() << f_G + f_B, m_r_G.cross(f_G) + m_r_B.cross(f_B))
+      .finished();
+}
+
+void QuaternionPIDController::init(const double W, const double B,
+                                   const Eigen::Vector3d &r_G,
+                                   const Eigen::Vector3d &r_B) {
+  m_W = W;
+  m_B = B;
+  m_r_G = r_G;
+  m_r_B = r_B;
+}
+
+void QuaternionPIDController::update_gain(Eigen::Vector6d p_gain,
+                                          Eigen::Vector6d i_gain,
+                                          Eigen::Vector6d d_gain) {
+  m_p_gain = p_gain;
+  m_i_gain.diagonal() << i_gain;
+  for (int i = 0; i < 6; i++) {
+    if (i_gain(i) == 0) {
+      integral(i) = 0;
+    }
+  }
```

Comment:
It's not immediately clear why the for loop is required here?

Reply:
Set as TODO